ipld/hamt/src/pointer.rs

```diff
@@ -81,26 +133,57 @@ where
 }
 
 /// Deserialize the Pointer like an untagged enum.
-impl<'de, K, V, H> Deserialize<'de> for Pointer<K, V, H>
+impl<'de, K, V, Ver, H> Deserialize<'de> for Pointer<K, V, Ver, H>
 where
     K: DeserializeOwned,
     V: DeserializeOwned,
+    Ver: self::version::Version,
 {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        Ipld::deserialize(deserializer).and_then(|ipld| ipld.try_into().map_err(de::Error::custom))
+        match Ver::NUMBER {
+            0 => {
+                let ipld = Ipld::deserialize(deserializer)?;
+                let (_key, value) = match ipld {
+                    Ipld::Map(map) => map
+                        .into_iter()
+                        .next()
+                        .ok_or("Expected at least one element".to_string()),
+                    other => Err(format!("Expected `Ipld::Map`, got {:#?}", other)),
+                }
```

Comment:
We should decode this as a tagged enum instead of using `Ipld` and looking at the types.

```rust
#[derive(Deserialize, Serialize)]
enum PointerDe<K, V> {
    #[serde(name = "l")]
    Link(Cid),
    #[serde(name = "v")]
    Values(KeyValuePair<K, V>),
}
```

Reply:
i had to change l and v to 0 and 1 respectively to read genesis header in forest. 
Changes made: 
```
    #[derive(Serialize)]
    pub(super) enum PointerSer<'a, K, V> {
        #[serde(rename = "0")]
        Vals(&'a [KeyValuePair<K, V>]),
        #[serde(rename = "1")]
        Link(&'a Cid),
    }

    #[derive(Deserialize, Serialize)]
    pub enum PointerDe<K, V> {
        #[serde(rename = "0")]
        Link(Cid),
        #[serde(rename = "1")]
        Vals(Vec<KeyValuePair<K, V>>),
    }
```
But ipld_hamt_v0 tests fail with:
```
thread 'test_default::test_hamtv0' panicked at 'called `Result::unwrap()` on an `Err` value: Dynamic(Serialization error for Cbor protocol: TypeMismatch { name: "CBOR tag", byte: 1 })', ipld/hamt/tests/hamt_tests.rs:1022:69
```

Reply:
Huh. I thought it was k/v. I must be thinking of IPFS HAMTs.

> But Hamtv0 tests in this repo fails with:

Hm. Something funky is going on there. Can you push the code?

Reply:
(or push it to a new branch)

Reply:
I've pushed the changes with failing hamtv0 tests.

Reply:
It was a sneaky bug. The Vals/Pointer variants were swapped.

Reply:
Ah i see, interesting! thanks